core/src/main/java/org/apache/accumulo/core/file/rfile/RFile.java

```diff
@@ -386,13 +392,23 @@ public void flushIfNeeded() throws IOException {
 
     private SampleLocalityGroupWriter sample;
 
-    LocalityGroupWriter(BlockFileWriter fileWriter, long blockSize, LocalityGroupMetadata currentLocalityGroup, SampleLocalityGroupWriter sample) {
+    private SummaryStatistics keyLenStats = new SummaryStatistics();
+    private double avergageKeySize = 0;
+
+    LocalityGroupWriter(BlockFileWriter fileWriter, long blockSize, long maxBlockSize, LocalityGroupMetadata currentLocalityGroup,
+        SampleLocalityGroupWriter sample) {
       this.fileWriter = fileWriter;
       this.blockSize = blockSize;
+      this.maxBlockSize = maxBlockSize;
       this.currentLocalityGroup = currentLocalityGroup;
       this.sample = sample;
     }
 
+    private boolean isGiantKey(Key k) {
+      // consider a key thats more than 3 standard deviations from previously seen key sizes as giant
+      return k.getSize() > keyLenStats.getMean() + keyLenStats.getStandardDeviation() * 3;
```

Comment:
TODO ensure there is a test for case when all keys are exactly the same size.  Test should assert more than one data block.